pkg/virt-operator/kubevirt.go

```diff
@@ -0,0 +1,527 @@
+	cache.WaitForCacheSync(stopCh, c.informers.Crd.HasSynced)
+	cache.WaitForCacheSync(stopCh, c.informers.Service.HasSynced)
+	cache.WaitForCacheSync(stopCh, c.informers.Deployment.HasSynced)
+	cache.WaitForCacheSync(stopCh, c.informers.DaemonSet.HasSynced)
+
+	// Start the actual work
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	log.Log.Info("Stopping KubeVirt controller.")
+}
+
+func (c *KubeVirtController) runWorker() {
+	for c.Execute() {
+	}
+}
+
+func (c *KubeVirtController) Execute() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+	err := c.execute(key.(string))
+
+	if err != nil {
+		log.Log.Reason(err).Infof("reenqueuing KubeVirt %v", key)
+		c.queue.AddRateLimited(key)
+	} else {
+		log.Log.V(4).Infof("processed KubeVirt %v", key)
+		c.queue.Forget(key)
+	}
+	return true
+}
+
+func (c *KubeVirtController) execute(key string) error {
+
+	// Fetch the latest KubeVirt from cache
+	obj, exists, err := c.kubeVirtInformer.GetStore().GetByKey(key)
+
+	if err != nil {
+		return err
+	}
+
+	if !exists {
+		// when the resource is gone, deletion was handled already
+		log.Log.Infof("KubeVirt resource not found")
+		c.kubeVirtExpectations.DeleteExpectations(key)
+		return nil
+	}
+
+	kv := obj.(*v1.KubeVirt)
+	logger := log.Log.Object(kv)
+
+	// If we can't extract the key we can't do anything
+	_, err = controller.KeyFunc(kv)
+	if err != nil {
+		log.Log.Reason(err).Errorf("Could not extract the key from the custom resource, will do nothing and not requeue.")
+		return nil
+	}
+
+	logger.Infof("Handling KubeVirt resource")
+
+	// only process the kubevirt deployment if all expectations are satisfied.
+	needsSync := c.kubeVirtExpectations.SatisfiedExpectations(key)
+	if !needsSync {
+		logger.Infof("Waiting for expectations to be fulfilled")
+		return nil
+	}
+
+	// Adds of all types are not done in one go. We need to set an expectation of 0 so that we can add something
+	c.kubeVirtExpectations.ResetExpectations(key)
+	if kv.DeletionTimestamp != nil {
+
+		log.Log.Info("Handling deletion")
+
+		// delete
+		if kv.Status.Phase == v1.KubeVirtPhaseDeleted {
+			log.Log.Info("Is already deleted")
+			return nil
+		}
+
+		// set phase to deleting
+		err = util.UpdatePhase(kv, v1.KubeVirtPhaseDeleting, c.clientset)
+		if err != nil {
+			log.Log.Errorf("Failed to update phase: %v", err)
+			return err
+		}
+
+		err := deletion.Delete(kv, c.clientset, c.stores, &c.kubeVirtExpectations)
+		if err != nil {
+			// deletion failed
+			err := util.UpdateCondition(kv, v1.KubeVirtConditionSynchronized, k8sv1.ConditionFalse, ConditionReasonDeletionFailedError, fmt.Sprintf("An error occurred during deletion: %v", err), c.clientset)
+			if err != nil {
+				log.Log.Errorf("Failed to set condition: %v", err)
+			}
+			return err
+		}
+
+		if c.stores.AllEmpty() {
+			// deletion successful
+			err = util.UpdatePhase(kv, v1.KubeVirtPhaseDeleted, c.clientset)
```

Comment:
We need to perform one single update here on the kubeVirt object, or else we may end up in a weird unrecoverable situations.


For example, If this update that sets the phase to Deletion works, and somehow the finalizer Patch fails, then the finalizer will never get removed... There's an if statement that exists early when phase == deleted.